File: src/livekit_agent_simulator/persona_traits.py
"""Portable caller trait library — expands short tags into system-prompt instructions.

Scenarios use ``Persona.spec.traits: ["impatient", "quiet", ...]``. Unknown tags are
passed through as free-text behavior hints (no consumer-specific coupling).
"""

from __future__ import annotations

# Canonical trait id → instruction bullets for the sim caller (English; agent speaks in locale).
TRAIT_LIBRARY: dict[str, list[str]] = {
    "polite": [
        "Stay courteous; use soft openers and thank the agent when appropriate.",
    ],
    "impatient": [
        "You are short on time: speak a bit faster, push for a quick answer,",
        "and do not tolerate long monologues.",
    ],
    "interrupts": [
        "You sometimes cut in while the agent is still talking with a brief correction",
        "or urgency (one short phrase). Do not monologue over them.",
    ],
    "quiet": [
        "You are reserved: leave pauses before answering; replies are very short.",
    ],
    "silent": [
        "You often go quiet after the agent speaks; only answer when necessary.",
    ],
    "confused": [
        "You misunderstand details once or twice and ask the agent to repeat simply.",
    ],
    "elderly": [
        "Speak slightly slower and ask the agent to speak clearly; prefer simple words.",
    ],
    "angry": [
        "You are mildly frustrated (not abusive). Express annoyance briefly if the agent",
        "is unclear, but stay on topic.",
    ],
    "fast_speaker": [
        "Speak quickly and denser than average phone speech; still 1–2 sentences max.",
    ],
    "soft_spoken": [
        "Keep volume soft and wording gentle; avoid sharp interruptions.",
    ],
    "non_native": [
        "You are fluent enough but occasionally ask for clarification of complex words.",
    ],
    "skeptical": [
        "Question vague claims; ask for concrete next steps or confirmation.",
    ],
    "chatty": [
        "You add a small extra detail about your situation, but still finish goals.",
    ],
    "backchannel": [
        "Occasionally acknowledge with a very short uh-huh / okay while listening,",
        "without stealing the full turn.",
    ],
    "hangup_threat": [
        "If the agent is unhelpful or loops the menu, you may briefly threaten to hang up,",
        "then give one more chance before ending the call.",
    ],
    "code_switch": [
        "You mainly use the call language but may mix one short English phrase if stuck.",
    ],
    "urgent": [
        "State urgency early; prefer concrete next steps over long explanations.",
    ],
}


def expand_traits(traits: list[str] | tuple[str, ...] | str) -> list[str]:
    """Return prompt lines for known traits + passthrough for unknown tags."""
    if isinstance(traits, str):
        traits = [traits]
    lines: list[str] = []
    seen: set[str] = set()
    unknown: list[str] = []
    for raw in traits:
        tag = str(raw).strip()
        if not tag:
            continue
        key = tag.lower().replace(" ", "_").replace("-", "_")
        if key in seen:
            continue
        seen.add(key)
        if key in TRAIT_LIBRARY:
            lines.extend(TRAIT_LIBRARY[key])
        else:
            unknown.append(tag)
    if unknown:
        lines.append(
            "Additional caller behavior (follow naturally): " + ", ".join(unknown) + "."
        )
    return lines


def list_trait_ids() -> list[str]:
    return sorted(TRAIT_LIBRARY.keys())

File: src/livekit_agent_simulator_rust/crates/lks-core/src/persona_traits.rs
//! Portable caller trait library — byte-parity port of `persona_traits.py`.
//!
//! Scenarios use `Persona.spec.traits: ["impatient", "quiet", ...]`. Unknown tags
//! pass through as free-text behavior hints. Length-band aliases (quiet/silent/
//! terse → quiet, chatty → chatty) resolve in CallerPolicyContext.resolved_verbosity.

use serde_json::Value as Json;

/// Canonical trait id → instruction bullets for the sim caller (English).
pub fn trait_library() -> std::collections::BTreeMap<&'static str, &'static [&'static str]> {
    let mut m = std::collections::BTreeMap::new();
    m.insert(
        "polite",
        &["Stay courteous; use soft openers and thank the agent when appropriate."][..],
    );
    m.insert(
        "impatient",
        &[
            "You are short on time: speak a bit faster, push for a quick answer,",
            "and do not tolerate long monologues.",
        ][..],
    );
    m.insert(
        "interrupts",
        &[
            "You sometimes cut in while the agent is still talking with a brief correction",
            "or urgency (one short phrase). Do not monologue over them.",
        ][..],
    );
    m.insert(
        "quiet",
        &["You are reserved: leave pauses before answering; replies are very short."][..],
    );
    m.insert(
        "terse",
        &["Keep answers minimal: one short clause when possible; no extra detail."][..],
    );
    m.insert(
        "silent",
        &["You often go quiet after the agent speaks; only answer when necessary."][..],
    );
    m.insert(
        "confused",
        &["You misunderstand details once or twice and ask the agent to repeat simply."][..],
    );
    m.insert(
        "elderly",
        &["Speak slightly slower and ask the agent to speak clearly; prefer simple words."][..],
    );
    m.insert(
        "angry",
        &[
            "You are mildly frustrated (not abusive). Express annoyance briefly if the agent",
            "is unclear, but stay on topic.",
        ][..],
    );
    m.insert(
        "fast_speaker",
        &[
            "Speak quickly and denser than average phone speech; stay within the length band above.",
        ][..],
    );
    m.insert(
        "soft_spoken",
        &["Keep volume soft and wording gentle; avoid sharp interruptions."][..],
    );
    m.insert(
        "non_native",
        &["You are fluent enough but occasionally ask for clarification of complex words."][..],
    );
    m.insert(
        "skeptical",
        &["Question vague claims; ask for concrete next steps or confirmation."][..],
    );
    m.insert(
        "chatty",
        &["You add a small extra detail about your situation, but still finish goals."][..],
    );
    m.insert(
        "backchannel",
        &[
            "Occasionally acknowledge with a very short uh-huh / okay while listening,",
            "without stealing the full turn.",
        ][..],
    );
    m.insert(
        "hangup_threat",
        &[
            "If the agent is unhelpful or loops the menu, you may briefly threaten to hang up,",
            "then give one more chance before ending the call.",
        ][..],
    );
    m.insert(
        "code_switch",
        &["You mainly use the call language but may mix one short English phrase if stuck."][..],
    );
    m.insert(
        "urgent",
        &["State urgency early; prefer concrete next steps over long explanations."][..],
    );
    m
}

/// Return prompt lines for known traits + passthrough for unknown tags.
/// traits may be a JSON array of strings or a single string.
pub fn expand_traits(traits: &Json) -> Vec<String> {
    let raw_items: Vec<String> = match traits {
        Json::String(s) => vec![s.clone()],
        Json::Array(a) => a
            .iter()
            .map(|v| match v {
                Json::String(s) => s.clone(),
                other => other.to_string(),
            })
            .collect(),
        _ => Vec::new(),
    };
    let lib = trait_library();
    let mut lines: Vec<String> = Vec::new();
    let mut seen: std::collections::HashSet<String> = std::collections::HashSet::new();
    let mut unknown: Vec<String> = Vec::new();
    for raw in &raw_items {
        let tag = raw.trim();
        if tag.is_empty() {
            continue;
        }
        let key = tag.to_lowercase().replace([' ', '-'], "_");
        if seen.contains(&key) {
            continue;
        }
        seen.insert(key.clone());
        if let Some(bullets) = lib.get(key.as_str()) {
            lines.extend(bullets.iter().map(|s| s.to_string()));
        } else {
            unknown.push(tag.to_string());
        }
    }
    if !unknown.is_empty() {
        lines.push(format!(
            "Additional caller behavior (follow naturally): {}.",
            unknown.join(", ")
        ));
    }
    lines
}

/// Sorted trait ids.
pub fn list_trait_ids() -> Vec<&'static str> {
    trait_library().keys().copied().collect()
}
